src/event-replay/reverse-file-stream.ts

```diff
@@ -0,0 +1,47 @@
+import * as stream from 'stream';
+import * as fs from 'fs';
+
+/**
+ * Streams lines from a text file in reverse, starting from the end of the file.
+ * Modernized version of https://www.npmjs.com/package/fs-reverse
+ */
+export class ReverseFileStream extends stream.Readable {
+  private fileDescriptor: number;
+  private position: number;
+
+  private lineBuffer: string[] = [];
+  private remainder: string = '';
+
+  constructor(filePath: string) {
+    super();
+    this.position = fs.statSync(filePath).size;
+    this.fileDescriptor = fs.openSync(filePath, 'r', 0o666);
+  }
+
+  _read(size: number): void {
+    while (this.lineBuffer.length === 0 && this.position > 0) {
+      // Read `size` bytes from the end of the file.
+      const length = Math.min(size, this.position);
+      const buffer = Buffer.alloc(length);
+      this.position = this.position - length;
+      fs.readSync(this.fileDescriptor, buffer, 0, length, this.position);
+
+      // Split into lines to fill the `lineBuffer`.
+      this.remainder = buffer + this.remainder;
```

Comment:
So this `[type Buffer] + [type string]` I think works because implicit `Buffer.toString([encoding])` is called, and the default encoding is utf8, but probably better to be explicit. I think normally typescript/linters would error here but we must have some rule misconfigured or disabled.